NEXUS-4968: Proper fix.

cstamas opened the pull request:
In short, not using "millisecond timestamp" for change detection.

In old code, the timestamp was not used at all to get the date
(except in one DEBUG log, but that was irrelevant, as change
date will be same as timestamp of logline it sits on!). The pure
reason of the timestamp was to support "change detection", but
it was wrong, as millisecond resolution is too coarse: think
concurrent changes, or just like in related issue, UT issues
with Windows running on Hypervisor and sleeping less than OS
slice and having JRE "under sleeping". See links in the
issue comments.

Also, there is a "funkiness" of the DefaultFSLocalRepositoryStorage class
with regard of mixing Plexus and jsr330 (constructor injection in plexus
was never supported), I did not touch that, but did remove the wrong
Map injection as that map is not a component have to be modifiable.

cstamas committed:
NEXUS-4968: Proper fix.

In short, not using "millisecond timestamp" for change detection.

In old code, the timestamp was not used at all to get the date
(except in one DEBUG log, but that was irrelevant, as change
date will be same as timestamp of logline it sits on!). The pure
reason of the timestamp was to support "change detection", but
it was wrong, as millisecond resolution is too coarse: think
concurrent changes, or just like in related issue, UT issues
with Windows running on Hypervisor and sleeping less than OS
slice and having JRE "under sleeping". See links in the
issue comments.

jdillon commented on nexus/nexus-configuration/src/test/java/org/sonatype/nexus/configuration/application/SimpleRemoteStorageContext.java:
```diff
@@ -26,11 +26,20 @@
+    @Override
+    public int getGeneration()
+    {
+        // TODO Auto-generated method stub
```
What is all this "TODO Auto-generated method stub" for?

cstamas replied:
The "Simple..." muck is removed.

jdillon commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/storage/AbstractStorageContext.java:
```diff
@@ -60,7 +87,9 @@ public StorageContext getParentStorageContext()
-        this.parent = parent;
+        // noop
+        getLogger().warn(
+            "Class {} uses illegal method invocation, org.sonatype.nexus.proxy.storage.AbstractStorageContext.setParentStorageContext( StorageContext ), please update the code!" );
```
Use of {} but no argument

cstamas replied:
Fixed, thanks.

jdillon commented on nexus/nexus-proxy/src/test/java/org/sonatype/nexus/proxy/storage/remote/commonshttpclient/HttpClientProxyUtilTest.java:
```diff
@@ -153,7 +180,7 @@ public String getHostname()
         long lastChanged = ctx.getLastChanged();
         // making sure we will have subtraction result grater than 0
-        Thread.sleep( 5 );
+        Thread.sleep( 15 );
```
Tests should strive to avoid use of time based mechanism to validate functionality.

cstamas replied:
Agreed, but this is the same test that exists in master (also, see [Simple fix](https://github.com/sonatype/nexus/pull/362) with same change). I just left the test in place -- just like the method call but made it deprecated.

cstamas committed:
NEXUS-4968: Added argument to log message (was missing)

cstamas committed:
NEXUS-4968: Reshuffle classes to make things simpler

The "Default..." context implementation are moved "forward"
to make all the "Simple..." class use unneeded.

peterlynch commented on nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/storage/AbstractStorageContext.java:
```diff
@@ -25,32 +26,58 @@
+        return generation.get();
+    }
+
+    @Deprecated
```
Why?

cstamas replied:
Left only to be backward compatible (Nexus Core is not using this field, but some plugin might).

Change detection should be done based on "generation" not on "change timestamp" (this is millisecond resolution)

peterlynch replied:
useful to add as a comment in source code?

peterlynch commented on nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/storage/AbstractStorageContext.java:
```diff
@@ -25,32 +26,58 @@
+    private AtomicInteger generation;
 
-    public AbstractStorageContext( StorageContext parent )
+    @Deprecated
```
Why? strange to deprecate a private member so you probably have a good reason ;)

cstamas replied:
To have a visual feedback in IDE, that for change detection this field should not be used.

Change detection based on milliseconds are bad. Thats why we need to sleep in UTs, but also think concurrent config changes. Using millisecond based change detection omits to "detect" changes happened in same moment and/or concurrently.

jdillon commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/storage/remote/AbstractRemoteRepositoryStorage.java:
```diff
@@ -122,10 +122,8 @@ protected synchronized RemoteStorageContext getRemoteStorageContext( ProxyReposi
-                            new Object[] { ctx, new Date( ctx.getLastChanged() ),
-                                new Date( repositoryContexts.get( repository.getId() ) ) } );
+                        getLogger().debug( "Remote context {} has been changed. Previous generation {}",
+                            new Object[] { ctx, repositoryContexts.get( repository.getId() ) } );
```
Should use 2 argument method now that only 2 {} exist (damn slf4j for not making these varargs already).

jdillon commented:
What does CI say?

cstamas commented:
CI scheduled:
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/297/

cstamas commented:
CI voted +1

peterlynch commented:
+1

adreghiciu commented:
+1
